code/modules/recycling/disposal/pipe.dm

```diff
@@ -40,85 +47,101 @@
 			dpdir |= turn(dir, -90)
 		if(initialize_dirs & DISP_DIR_FLIP)
 			dpdir |= turn(dir, 180)
+		if(initialize_dirs & DISP_DIR_UP)
+			dpdir |= UP
+		if(initialize_dirs & DISP_DIR_DOWN)
+			dpdir |= DOWN
 
-	AddElement(/datum/element/undertile, TRAIT_T_RAY_VISIBLE)
+	//AddElement(/datum/element/undertile, TRAIT_T_RAY_VISIBLE, nullspace_target = TRUE)
+	AddComponent(/datum/component/nullspace_undertile, invisibility_trait = TRAIT_T_RAY_VISIBLE, nullspace_when_underfloor_visible = FALSE)
 
 // pipe is deleted
 // ensure if holder is present, it is expelled
 /obj/structure/disposalpipe/Destroy()
-	var/obj/structure/disposalholder/H = locate() in src
-	if(H)
-		H.active = FALSE
-		expel(H, get_turf(src), 0)
+	var/obj/structure/disposalholder/holder = locate() in associated_loc
+	if(holder)
+		holder.active = FALSE
+		expel(holder, associated_loc, 0)
 	stored = null //The qdel is handled in expel()
+
+	if(associated_loc)
```

Comment:
I really do wonder if this shouldn't somehow be a property of the component

Reply:
now a property of the element